### ISSUE TITLE
warpctc install error: g++: error: unrecognized command line option ‘-R’

### ISSUE THREAD

**Issue by bmharper** [SeanNaren/deepspeech.pytorch#159]
I'm trying to build and install warpctc (https://github.com/SeanNaren/warp-ctc) from source, but I'm stumped at the error:

   g++: error: unrecognized command line option ‘-R’

This happens when I run `python setup.py install` from `pytorch_bindings`.

Apologies for posting to this repo, but I'm unable to open an issue on the warp-ctc fork.

My environment is:

    ubuntu 16.04
    CPU only (no CUDA)
    anaconda 5.0.0.1
    apt install g++ cmake
    conda install numpy pyyaml mkl setuptools cffi
    Install pytorch from source - OK
    Build libwarpctc from source - OK
    cd pytorch_binding && python setup.py install - DIE with -R option

My GCC is version 5.4.0.
I've also tried GCC 4.8, but I get the same error.
I'm really stuck on this, since it seems like the -R is being generated by CFFI, but I'm hesitant to go deep into that code to understand why this pops up in the first place. As far as I can tell, the setup I've got here is pretty standard.

I've reinstalled this machine from scratch, tonight, and done *nothing* on it except for try to get warpctc compiled.

Here is the full gcc command line:

```
/usr/bin/gcc-4.8 -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/home/ben/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include -I/home/ben/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include/TH -I/home/ben/dev/warp-ctc/include -I/home/ben/anaconda3/include/python3.6m -c /home/ben/dev/warp-ctc/pytorch_binding/src/binding.cpp -o build/temp.linux-x86_64-3.6/home/ben/dev/warp-ctc/pytorch_binding/src/binding.o -std=c++11 -fPIC
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++ [enabled by default]
/usr/bin/g++-4.8 -pthread -shared -B /home/ben/anaconda3/compiler_compat -L/home/ben/anaconda3/lib -Wl,-rpath=/home/ben/anaconda3/lib,--no-as-needed build/temp.linux-x86_64-3.6/build/_warp_ctc.o build/temp.linux-x86_64-3.6/home/ben/dev/warp-ctc/pytorch_binding/src/binding.o -L/home/ben/dev/warp-ctc/build -L/home/ben/anaconda3/lib -R/home/ben/dev/warp-ctc/build -lwarpctc -lpython3.6m -o build/lib.linux-x86_64-3.6/warpctc_pytorch/_warp_ctc.cpython-36m-x86_64-linux-gnu.so
g++-4.8: error: unrecognized command line option ‘-R’
error: command '/usr/bin/g++-4.8' failed with exit status 1
```

**Comment by bmharper** [comment]
So... for posterity, I have managed to work around this issue with an uber hack:
I created the following script:

```python
#!/home/ben/anaconda3/bin/python3

import sys
import os

args = ""
for i in range(0, len(sys.argv)):
    if i == 0:
        continue
    arg = sys.argv[i]
    if arg[:2] != "-R":
        args += '"' + arg + '" '

os.system("gcc " + args)
```
This script discards any argument starting with `-R`.

I then set this script as my compiler, via
`export CC=/home/ben/clean-gcc`
Also, I make a copy of it for g++, replacing the `os.system` call to `gcc` with `g++`, and set the CXX environment variable accordingly:
`export CXX=/home/ben/clean-g++`

Then, `python setup.py install` of warpcpc_pytorch succeeds.

**Comment by slbinilkumar** [comment]
pls try with this
For _R option use 
 -Wl,-rpath

**Comment by bmharper** [comment]
Hi @slbinilkumar,
Do you have any idea how one could instruct the cffi build system to make that change?

**Comment by slbinilkumar** [comment]
I had done like this - copy the command and change -R with -Wl,-rpath

/usr/bin/g++ -pthread -shared -B /root/anaconda3/compiler_compat -L/root/anaconda3/lib -Wl,-rpath=/root/anaconda3/lib,--no-as-needed build/temp.linux-x86_64-3.6/build/_warp_ctc.o build/temp.linux-x86_64-3.6/data/deepspeech/pytorch_instalation/NC12/warp-ctc/pytorch_binding/src/binding.o -L/data/deepspeech/pytorch_instalation/NC12/warp-ctc/build -L/root/anaconda3/lib -Wl,-rpath /data/deepspeech/pytorch_instalation/NC12/warp-ctc/build -lwarpctc -lpython3.6m -o build/lib.linux-x86_64-3.6/warpctc_pytorch/_warp_ctc.cpython-36m-x86_64-linux-gnu.so

**Comment by SeanNaren** [comment]
I'm not sure how to fix this... any suggestions from anyone more experienced than myself?

**Comment by jxlijunhao** [comment]
following the above solutions and some other tips, solved my problem in Ubuntu16.04 and CUDA9.0,

>   Install pytorch audio, beause  pytorch_binding/warp_ctc/__init__.py include, 
  from ._warp_ctc import lib as _lib, ffi as _ffi, so we need to install pytorch audio:
> sudo apt-get install sox libsox-dev libsox-fmt-all
git clone https://github.com/pytorch/audio.git
cd audio
pip install cffi
python setup.py install

> git clone https://github.com/SeanNaren/warp-ctc.git
cd warp-ctc
mkdir build; cd build
cmake ..
make
export CUDA_HOME="/usr/local/cuda"
cd ../pytorch_binding
python setup.py install

then following  @slbinilkumar  ,  replace R with -Wl,-rpath the error  g++ commd,

and test:

`python`
import torch
    from torch.autograd import Variable
    from warpctc_pytorch import CTCLoss
    ctc_loss = CTCLoss()
    # expected shape of seqLength x batchSize x alphabet_size
    probs = torch.FloatTensor([[[0.1, 0.6, 0.1, 0.1, 0.1], [0.1, 0.1, 0.6, 0.1, 0.1]]]).transpose(0, 1).contiguous()
    labels = Variable(torch.IntTensor([1, 2]))
    label_sizes = Variable(torch.IntTensor([2]))
    probs_sizes = Variable(torch.IntTensor([2]))
    probs = Variable(probs, requires_grad=True) # tells autograd to compute gradients for probs
    cost = ctc_loss(probs, labels, probs_sizes, label_sizes)
    cost.backward()

**Comment by SeanNaren** [comment]
Thanks guys, working on trying to fix the external libraries causing the issues.
